src/Kestrel.Transport.Sockets/Internal/SocketAwaitableEventArgs.cs

```diff
@@ -5,74 +5,137 @@
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
 {
-    public class SocketAwaitableEventArgs : SocketAsyncEventArgs, ICriticalNotifyCompletion
+    public class SocketAwaitableEventArgs : SocketAsyncEventArgs, IValueTaskSource<int>
     {
-        private static readonly Action _callbackCompleted = () => { };
+        private static readonly Action<object> _callbackCompleted = _ => { };
 
         private readonly PipeScheduler _ioScheduler;
 
-        private Action _callback;
+        private Action<object> _completion;
 
         public SocketAwaitableEventArgs(PipeScheduler ioScheduler)
         {
             _ioScheduler = ioScheduler;
         }
 
-        public SocketAwaitableEventArgs GetAwaiter() => this;
-        public bool IsCompleted => ReferenceEquals(_callback, _callbackCompleted);
+        protected override void OnCompleted(SocketAsyncEventArgs _)
+        {
+            Action<object> completion = _completion;
+
+            if (completion != null || (completion = Interlocked.CompareExchange(ref _completion, _callbackCompleted, null)) != null)
+            {
+                Debug.Assert(!ReferenceEquals(completion, _callbackCompleted));
+
+                object state = UserToken;
+                UserToken = null;
 
-        public int GetResult()
+                _ioScheduler.Schedule(completion, state);
+            }
+        }
+
+        public ValueTaskSourceStatus GetStatus(short token)
         {
-            Debug.Assert(ReferenceEquals(_callback, _callbackCompleted));
+            if (ReferenceEquals(_completion, _callbackCompleted))
+            {
+                if (SocketError != SocketError.Success)
+                {
+                    return ValueTaskSourceStatus.Faulted;
+                }
 
-            _callback = null;
+                return ValueTaskSourceStatus.Succeeded;
+            }
+
+            return ValueTaskSourceStatus.Pending;
+        }
+
+        public void OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags)
+        {
+            // We're ignoring ValueTaskSourceOnCompletedFlags and token since this is used in a single place in the code base
+            // so we don't need to handle capturing the SynchronizationContext and ExecutionContext
+            Debug.Assert((flags & ValueTaskSourceOnCompletedFlags.FlowExecutionContext) == 0, "FlowExecutionContext is set.");
+
+            // Use UserToken to carry the continuation state around
+            UserToken = state;
+            Action<object> awaitableState = Interlocked.CompareExchange(ref _completion, continuation, null);
 
-            if (SocketError != SocketError.Success)
+            if (ReferenceEquals(awaitableState, _callbackCompleted))
+            {
+#if NETCOREAPP2_1
+                ThreadPool.QueueUserWorkItem(continuation, state, preferLocal: false);
+#else
+                Task.Factory.StartNew(continuation, state);
+#endif
```

Comment:
Since you null out UserToken in `OnCompleted(SocketAsyncEventArgs` it's probably best to also do that here.

Reply:
I'll null it out after the work item is invoked but before calling the continuation.

Reply:
You don't have to wait until after the work item invoked to null out UserToken for the same reason you don't need to in `OnCompleted(SocketAsyncEventArgs`. Just use the local state reference which isn't null. Technically you don't need to even set UserToken in this case, you could just set it in an else block at the end.